### PR TITLE
Review fixes of #872

### PR DESCRIPTION
**Pull request by Nityanand13**
[Review fixes of #872](https://github.com/mattermost/mattermost-plugin-jira/pull/872)

### CODE REVIEW

File: server/issue_test.go [hunk 1]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+		wh          webhook
+		expectedErr error
+	}{
+		"No Field": {
+			p: &Plugin{
+				instanceStore: mockInstanceStore{},
+				userStore:     mockUserStore{},
+			},
+			instanceID: testInstance1.InstanceID,
+			client:     nil,
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:333]: Here create more sub test cases like when client is nil and not nil similarly connection is nil and not nil

File: server/issue_test.go [hunk 2]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+	p.instanceStore = p.getMockInstanceStoreKV(1)
+	p.userStore = getMockUserStoreKV()
+
+	tests := map[string]struct {
+		wh         *webhook
+		instanceID types.ID
+		reporter   *jira.User
+	}{
+		"No reporter": {
+			wh: &webhook{
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:399]: this is same for both tc and instant id so create a constants for this

File: server/issue_test.go [hunk 3]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+			instanceID: testInstance1.InstanceID,
+			reporter:   &jira.User{},
+		},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			l1 := len(tt.wh.notifications)
+			p.applyReporterNotification(
+				tt.wh,
+				tt.instanceID,
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:454]: Here you can create more sub-test cases like when instance id nil think of more scnerios

File: server/issue_test.go [hunk 4]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+				},
+				notifications: []webhookUserNotification{},
+			},
+			instanceID: testInstance1.InstanceID,
+			reporter:   &jira.User{},
+		},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			l1 := len(tt.wh.notifications)
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:451]: give better name than l1

File: server/issue_test.go [hunk 5]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+		instanceID    types.ID
+		jiraAccountID string
+		jiraUsername  string
+		connection    *Connection
+		expectedErr   error
+	}{
+		"No instanceID": {
+			wh:            &webhook{},
+			instanceID:    "",
+			jiraAccountID: "",
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:489]: same comments as in above test cases

File: server/issue_test.go [hunk 6]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+func TestShouldReceiveNotification(t *testing.T) {
+	s := ConnectionSettings{}
+	falseValue := false
+	tests := map[string]struct {
+		role         string
+		notification bool
+	}{
+		subCommandAssignee: {
+			role:         subCommandAssignee,
+			notification: falseValue,
```
[the remark this answers is not in the document]
Comment (Nityanand13) [reply server/issue_test.go:292]: But earlier at most of the places they have used `falseValue` for false and `trueValue` for true

File: server/issue_test.go (hunk 2 above)
Comment (Nityanand13) [reply server/issue_test.go:399]: Now it is not same

File: server/issue_test.go (hunk 3 above)
Comment (Nityanand13) [reply server/issue_test.go:454]: We cannot initialize instanceID as `nil` because ` We cannot use nil as types.ID value in struct literal`, instead made an another tc in which in which instanceID is initialized with `"initialize"`

File: server/issue_test.go (hunk 1 above)
Comment (Nityanand13) [reply server/issue_test.go:333]: It  is already there, by the way added 1 more test case in this function.

File: server/issue_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kshitij-Katiyar) [reply server/issue_test.go]: because at that time they are using pointer but now you aren't

File: server/issue_test.go [hunk 7]
```diff
@@ -279,6 +280,265 @@ func TestRouteShareIssuePublicly(t *testing.T) {
 	}
 }
 
+func TestShouldReceiveNotification(t *testing.T) {
+	cs := ConnectionSettings{}
+	cs.RolesForDMNotification = make(map[string]bool)
+	cs.RolesForDMNotification[subCommandAssignee] = true
+	cs.RolesForDMNotification[subCommandMention] = true
+	falseValue := false
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:288]: there is no need for these variables use them directly

File: server/issue_test.go [hunk 8]
```diff
@@ -279,6 +280,265 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+			notification: falseValue,
+		},
+		subCommandWatching: {
+			role:         subCommandWatching,
+			notification: falseValue,
+		},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			val := cs.ShouldReceiveNotification(
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:314]:
```suggestion
			val := cs.ShouldReceiveNotification( tt.role )
```

File: server/issue_test.go [hunk 9]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+			wh: webhook{
+				JiraWebhook: &JiraWebhook{
+					Issue: jira.Issue{},
+				},
+			},
+			expectedErr: nil,
+		},
+		"Unable to load instance": {
+			instanceID: "instanceID",
+			client:     nil,
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:349]: both client and connection is nil in this case and the other as well have a test case where one is nil

File: server/issue_test.go [hunk 10]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+	}
+	tests := map[string]struct {
+		instanceID types.ID
+		reporter   *jira.User
+	}{
+		"Unable to load instance": {
+			instanceID: "instanceID",
+			reporter:   &jira.User{},
+		},
+		"No reporter": {
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:443]:
```suggestion
		"Reporter is nil": {
```

File: server/issue_test.go [hunk 11]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			totalNotifications := len(wh.notifications)
+			p.applyReporterNotification(
+				wh,
+				tt.instanceID,
+				tt.reporter,
+			)
+			if tt.reporter == nil || tt.instanceID == "instanceID" {
+				assert.Equal(t, len(wh.notifications)-totalNotifications, 0)
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:461]:
```suggestion
				assert.Equal(t, len(wh.notifications), totalNotifications)
```

File: server/issue_test.go [hunk 12]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+			} else {
+				assert.Equal(t, len(wh.notifications)-totalNotifications, 1)
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:463]:
```suggestion
				assert.Equal(t, len(wh.notifications), 1+totalNotifications)
```

File: server/issue_test.go [hunk 13]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+			jiraAccountID: "",
+			jiraUsername:  "",
+			connection:    nil,
+			expectedErr:   errors.New("instance " + fmt.Sprintf("\"%s\"", "instanceID") + " not found"),
+		},
+		"Success": {
+			wh:            &webhook{},
+			instanceID:    testInstance1.InstanceID,
+			jiraAccountID: "",
+			jiraUsername:  "",
```
Comment (Kshitij-Katiyar) [inline server/issue_test.go:500]: both username and accountID is empty in all test cases make them constant and have them equal to mock value not empty

File: server/user_test.go [hunk 14]
```diff
@@ -20,23 +20,23 @@ func TestUserSettings_String(t *testing.T) {
 		"notifications on": {
 			settings: ConnectionSettings{
 				Notifications: valueTrue,
-				RolesForDMNotification: map[string]*bool{
-					subCommandAssignee: nil,
-					subCommandMention:  nil,
-					subCommandReporter: nil,
-					subCommandWatching: nil,
+				RolesForDMNotification: map[string]bool{
+					subCommandAssignee: valueTrue,
```
Comment (Kshitij-Katiyar) [inline server/user_test.go:24]: no need to have this variable

File: server/issue_test.go [hunk 15]
```diff
@@ -279,6 +280,255 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+					Issue: jira.Issue{
+						Fields: &jira.IssueFields{
+							Creator: &jira.User{},
+						},
+					},
+				},
+			},
+			expectedErr: errors.New("instance " + fmt.Sprintf("\"%s\"", "instanceID") + " not found"),
+		},
+		"Success": {
```
[the remark this answers is not in the document]
Comment (Nityanand13) [reply server/issue_test.go:362]: Ok..

File: server/user.go [hunk 16]
```diff
@@ -34,17 +34,17 @@ type Connection struct {
 	DefaultProjectKey  string `json:"default_project_key,omitempty"`
 }
 
-func (c *Connection) JiraAccountID() types.ID {
-	if c.AccountID != "" {
-		return types.ID(c.AccountID)
+func (connection *Connection) JiraAccountID() types.ID {
```
Comment (Kshitij-Katiyar) [inline server/user.go:37]: why this change i think the previous change also looks good to me

Comment (Nityanand13) [reply server/user.go:37]: Because of this only i was getting lint error: `ST1016: methods on the same type should have the same receiver name (seen 1x "c", 1x "connection") (stylecheck)`

File: server/kv.go [hunk 17]
```diff
@@ -568,13 +568,13 @@ func UpdateInstances(store InstanceStore, updatef func(instances *Instances) err
...
 // MigrateV2Instances migrates instance record(s) from the V2 data model.
 //
-// - v2keyKnownJiraInstances ("known_jira_instances") was stored as a
-//   map[string]string (InstanceID->Type), needs to be stored as Instances.
-//   https://github.com/mattermost/mattermost-plugin-jira/blob/885efe8eb70c92bcea64d1ced6e67710eda77b6e/server/kv.go#L375
-// - v2keyCurrentJIRAInstance ("current_jira_instance") stored an Instance; will
-//   be used to set the default instance.
-// - The instances themselves should be forward-compatible, including
-// 	 CurrentInstance.
+//   - v2keyKnownJiraInstances ("known_jira_instances") was stored as a
```
[the remark this answers is not in the document]
Comment (Nityanand13) [reply server/kv.go:571]: this i got after running gofmt

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/issue_test.go [hunk 18]
```diff
@@ -279,6 +280,238 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+		"Issue Field not found": {
+			instanceID: testInstance1.InstanceID,
+			client:     nil,
+			connection: nil,
+			wh: webhook{
+				JiraWebhook: &JiraWebhook{
+					Issue: jira.Issue{},
+				},
+			},
+			expectedErr: nil,
```
[the remark this answers is not in the document]
Comment (Nityanand13) [reply server/issue_test.go:372]: No, in this case `client`, `connection` and `error` all the three will be `nil`

**Review by manojmalik20 — APPROVED** [review]

File: server/kv.go (hunk 17 above)
Comment (javaguirre) [inline server/kv.go:571]: I think we can restore the changes in this file, there aren't any.

File: server/issue.go [hunk 19]
```diff
@@ -1133,13 +1133,38 @@ func (p *Plugin) GetUserSetting(wh *webhook, instanceID types.ID, jiraAccountID,
...
 	if val, ok := s.RolesForDMNotification[role]; ok {
-		return *val
+		return val
 	}
 
 	// Check old setting for backwards compatibility
 	return s.Notifications
 }
 
+func (p *Plugin) fetchConnectedUserFromAccount(account map[string]string, instance Instance) (Client, *Connection, error) {
```
Comment (javaguirre) [inline server/issue.go:1143]: Why is this needed?

File: server/http_test.go [hunk 20]
```diff
@@ -316,9 +316,9 @@ func TestSubscribe(t *testing.T) {
 			api := &plugintest.API{}
 			p := Plugin{}
 
-			api.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...).Return(nil)
-			api.On("LogError", mockAnythingOfTypeBatch("string", 10)...).Return(nil)
-			api.On("LogError", mockAnythingOfTypeBatch("string", 13)...).Return(nil)
+			api.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...).Return()
```
Comment (javaguirre) [inline server/http_test.go:319]: why it was nil before and now it's not? It doesn't seem like a needed change, but I might be wrong.

**Review by javaguirre — CHANGES_REQUESTED** [review]
Thank you! just a small detail on a commented change and a couple of questions to get a bit more of context, let me know! 👍

**Comment by javaguirre** [comment]
Sorry, I didn't see I was in your fork, I will wait until you send the PR to the MM repo 👍

File: server/kv.go (hunk 17 above)
Comment (Nityanand13) [reply server/kv.go:571]: @javaguirre , I was getting this error from `server/kv.go`: `File is not gofmt -ed with -s` on running `make check-style` command because the file was not properly formatted according to Go rules. So I run `gofmt -s -w ./server` for formatting.

File: server/http_test.go (hunk 20 above)
Comment (Nityanand13) [reply server/http_test.go:319]: @javaguirre The function `LogError` and `LogDebug` does not have a return type.
